IPA experimental mm_host code sample (#1)
dancgr/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/ip_authoring_mmhost/README.md

# Memory-Mapped Interfaces (mmhost)

> This FPGA tutorial demonstrates how to use the memory-mapped interfaces with IP authoring and set the configurable parameters.

Reviewer: Is this only for when using IP authoring or will this feature also be available when using full board designs?

> ***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>

Reviewer: Replace this entire paragraph with:
```
### Additional Documentation
- [Explore SYCL* Through Intel&reg; FPGA Code Samples](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of FPGAs and SYCL.
- [FPGA Optimization Guide for Intel&reg; oneAPI Toolkits](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) helps you understand how to target FPGAs using SYCL and Intel&reg; oneAPI Toolkits.
- [Intel&reg; oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) helps you understand target-independent, SYCL-compliant programming using Intel&reg; oneAPI Toolkits.
```

and move it to just before "Key Concepts"

The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
The [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming.

| Optimized for                     | Description
---                                 |---

> | OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10

Reviewer: ```suggestion
| OS                                | Linux* Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
```

> | Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*

Reviewer: ```suggestion
| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg;10 GX FPGA <br> Intel&reg;FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg;10 SX) <br> Intel&reg;FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel&reg;FPGA PAC hardware is only compatible with Ubuntu 18.04*
```

> | Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit

Reviewer: ```suggestion
| Software                          | Intel&reg;oneAPI DPC++ Compiler <br> Intel&reg;FPGA Add-On for oneAPI Base Toolkit
```

> | What you will learn               |  The basic concepts of on-chip memory attributes <br> How to apply memory attributes in your program <br> How to confirm that the memory attributes were respected by the compiler <br> A case study of the type of performance/area trade-offs enabled by memory attributes

Reviewer: This description should be updated to the mmhost description

| Time to complete                  | 30 minutes

## Purpose

> With the mmhost interface controls, the user can configure memory-mapped interfaces for their components.

Reviewer: ```suggestion
This FPGA tutorial demonstrates how to make a component use memory-mapped interfaces and how to configure these interfaces.
```

> ### Memory-mapped interface parameters

Reviewer: I think it would be useful to quickly remind the readers what a memory-mapped interface is.

The following parameters are available for configuration:

> 

Reviewer: Please align the following table so that it looks nice when reading the raw readme as well

| Memory Attribute                 | Description
---                                |---

> | `buffer_location`          | The address space of the interface that associates with the host.

Reviewer: I'm not sure to understand this sentence, could you clarify? How can an interface have an address space? 
How can one know which number corresponds to which address space?

| `awidth`          | The width of the memory-mapped data bus in bits
| `dwidth`          | The width of the memory-mapped address bus in bits.

> | `latency`         | The guaranteed latency from when a read command exits the component when the external memory returns valid read data.

Reviewer: "guaranteed"  indicates to me that the latency may be variable, but the provided value is the upper bound. Is that what you intended? if not, maybe replace "guaranteed" with "exact"

| `readwrite_mode`  | The port direction of the interface. (0: Read & Write, 1: Read only, 2: Write only)

> | `maxburst`        | The maximum number of data transfers that can associate with a read or write transaction.

Reviewer: This is not very clear to me :(

| `align`           | The alignment of the base pointer address in bytes. 

> | `waitrequest`     | Adds the waitrequest signal that is asserted by the agent when it is unable to respond to a read or write request.

Reviewer: Is then "agent" a well known term?
Maybe this is all well known, but as a user who have no idea about about mmhost works, I don't really understand how all this works :)

### Default interface

> If no memory-mapped interface parameters is informed for a pointer, then a mmhost interface with all default parameters is assumed for it.

Reviewer: ```suggestion
If no memory-mapped interface parameters are informed for a pointer, then a mmhost interface with all default parameters is assumed for it.
```
Can you only specify a subset of parameters?

### Memory-mapped interface implementation restriction

> Memory-mapped interface is restricted to function implementations, which means that it is not supported for lambda implementation. Also, all parameters must be preset, even if just one parameter is being changed from the default values.

Reviewer: This second sentence answers my question above. I think it should go under the default interface section
Does it work with functions? or must it be in a functor?

> Memory-mapped interface is restricted to function implementations, which means that it is not supported for lambda implementation. Also, all parameters must be preset, even if just one parameter is being changed from the default values.

Reviewer: ```suggestion
Memory-mapped interface is restricted to function implementations, which means that it is not supported for lambda implementations. Also, all parameters must be preset, even if just one parameter is being changed from the default values.
```

> #### Example Functor

Reviewer: ```suggestion
#### Functor Example
```

```c++
struct MyIP {
  mmhost(
    ... // All mmhost parameters.
  ) int *my_pointer;

  MyIP(int *input_pointer) : my_pointer(input_pointer) { ... } // Constructor for our IP.
    ...
  }

  void operator()() const {
    ...
    // Functor code
    my_pointer[x] = y;
    ...
  }
}
```

#### Example 1: How to set-up a correct mmhost interface
```c++
// A memory-mapped interface must contain all parameters, in the following order:
  mmhost(

>     1,       // buffer_location or aspace

Reviewer: What does 1 correspond to?

    28,      // address width
    64,      // data width
    16,      // latency
    1,       // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
    1,       // maxburst
    0,       // align, 0 defaults to alignment of the type
    1        // waitrequest, 0: false, 1: true
  ) int *memory_mapped_pointer;
```

> #### Example 2: Changing default mmhost interface.

Reviewer: Does this mean that the previous example was using the default values? If so please make it clear

```c++
// A memory-mapped interface that is implemented as a register.
  register_map_mmhost(
    1,       // buffer_location or aspace
    28,      // address width
    64,      // data width
    16,      // latency
    1,       // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
    1,       // maxburst
    0,       // align, 0 defaults to alignment of the type
    1        // waitrequest, 0: false, 1: true
  ) int *memory_mapped_pointer;
// A memory-mapped interface that is implemented as a wire.
  conduit_mmhost(
    1,       // buffer_location or aspace
    28,      // address width
    64,      // data width
    16,      // latency
    1,       // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
    1,       // maxburst
    0,       // align, 0 defaults to alignment of the type
    1        // waitrequest, 0: false, 1: true
  ) int *memory_mapped_pointer;

> ```

Reviewer: These are identical :/

### Memory Base Addresses assigned by the compiler

> The compiler infers a unique global memory for each pointer argument that is annotated with a buffer location. An entry describing the interface of each such memory can be found in the file ipinterfaces.xml which is generated in the "*.prj" output directory. The entry specifies various properties of the memory including the start address that was assigned to the memory by the compiler. Each buffer_location will have its address defined there, among other parameters. 

Reviewer: "that is annotated with a buffer location" this is a comment from the compiler standpoint, rigth? Is the user supposed to know how a pointer is "annotated with a buffer location"?

> The compiler infers a unique global memory for each pointer argument that is annotated with a buffer location. An entry describing the interface of each such memory can be found in the file ipinterfaces.xml which is generated in the "*.prj" output directory. The entry specifies various properties of the memory including the start address that was assigned to the memory by the compiler. Each buffer_location will have its address defined there, among other parameters. 

Reviewer: ```suggestion
The compiler infers a unique global memory for each pointer argument that is annotated with a buffer location. An entry describing the interface of each of these memory can be found in the file `ipinterfaces.xml` which is generated in the "*.prj" output directory that is created when compiling for hardware or generating a report. Each entry specifies various properties of a memory including its start address that was assigned by the compiler. Each `buffer_location` will have its address defined there, among other parameters. 
```

> To understand how these start addresses are generated and how the compiler assigns special values to the top bits of the addresses please see the complete documentation on memory-mapped interfaces. If the parameters are not set correctly, this can lead to undefined behaviour.

Reviewer: Are you referring to the mmhost parameters described above? If so, please move this last sentence to when you discuss the setting of these parameters

> The file will have the following entries based on the kernel arguments used in this tutorial:Example of a pointer argument in the ipinterfaces.xml:

Reviewer: ```suggestion
Example of a pointer argument in the `ipinterfaces.xml` file based on the kernel arguments used in this tutorial:
```

```xml
<global_mem name="1" default="1" max_bandwidth="0" config_addr="0x0" type="device private" allocation_type="host,shared">
  <interface name="1" type="agent" width="64" address="0x2000000000000" latency="16" latency_type="fixed" waitrequest="1" size="0x10000000">
    <port name="mem1_r" direction="r"/>
  </interface>
</global_mem>
<global_mem name="2" max_bandwidth="0" config_addr="0x0" type="device private" allocation_type="host,shared">
  <interface name="2" port="mem2_rw" type="agent" width="64" address="0x3000000000000" latency="16" latency_type="fixed" waitrequest="1" size="0x10000000"/>
</global_mem>
<global_mem name="3" max_bandwidth="0" config_addr="0x0" type="device private" allocation_type="host,shared">
    <interface name="3" port="mem3_rw" type="agent" width="64" address="0x4000000000000" latency="16" latency_type="fixed" waitrequest="1" size="0x10000000"/>
</global_mem>

> ```

Reviewer: I don't see a `buffer_location` in this description

### Tutorial Code Overview

> This tutorial demonstrates how to implement a memory-mapped interface in an IP authoring flow. The design performs a multiply and add operation on two input vectors, stores the result of the multiple in the second vector, and the result of the add in an output-only vector.

Reviewer: ```suggestion
This tutorial demonstrates how to implement a memory-mapped interface in an IP authoring flow. The design performs a multiply and an add operation on two input vectors, stores the result of the multiple in the second vector, and the result of the add in an output-only vector.
```

> * Memory-mapped interfaces needs to contain all parameters to work correctly.

Reviewer: ```suggestion
* Memory-mapped interfaces needs to contain all parameters to work correctly.
```
Is this needed here?

> * Available parameters and default values can be fount at the header <sycl/ext/intel/experimental/interfaces.hpp>

Reviewer: This should be moved to the section discussing the parameters

> ## License

Reviewer: Please move this license section to be the very last section

Code samples are licensed under the MIT license. See
[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.

Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)

## Building the `mmhost` Tutorial

> 

Reviewer: ```suggestion
> **Note**: If you have not already done so, set up your CLI
> environment by sourcing  the `setvars` script located in
> the root of your oneAPI installation.
>
> Linux*:
> - For system wide installations: `. /opt/intel/oneapi/setvars.sh`
> - For private installations: `. ~/intel/oneapi/setvars.sh`
>
> Windows*:
> - `C:\Program Files(x86)\Intel\oneAPI\setvars.bat`
>
>For more information on environment variables, see **Use the setvars Script** for [Linux or macOS](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html), or [Windows](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
```

### Include Files
The included header `dpc_common.hpp` is located at `%ONEAPI_ROOT%\dev-utilities\latest\include` on your development system.

### Running Samples in DevCloud

> If running a sample in the Intel DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode. Compiles to FPGA are only supported on fpga_compile nodes. Executing programs on FPGA hardware is only supported on fpga_runtime nodes of the appropriate type, such as fpga_runtime:arria10 or fpga_runtime:stratix10.  Neither compiling nor executing programs on FPGA hardware are supported on the login nodes. For more information, see the Intel® oneAPI Base Toolkit Get Started Guide ([https://devcloud.intel.com/oneapi/documentation/base-toolkit/](https://devcloud.intel.com/oneapi/documentation/base-toolkit/)).

Reviewer: ```suggestion
If running a sample in the Intel&reg; DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode. Compiles to FPGA are only supported on fpga_compile nodes. Executing programs on FPGA hardware is only supported on fpga_runtime nodes of the appropriate type, such as fpga_runtime:arria10 or fpga_runtime:stratix10.  Neither compiling nor executing programs on FPGA hardware are supported on the login nodes. For more information, see the Intel&reg;oneAPI Base Toolkit Get Started Guide ([https://devcloud.intel.com/oneapi/documentation/base-toolkit/](https://devcloud.intel.com/oneapi/documentation/base-toolkit/)).
```

When compiling for FPGA hardware, it is recommended to increase the job timeout to 12h.

### Using Visual Studio Code*  (Optional)

> You can use Visual Studio Code (VS Code) extensions to set your environment, create launch configurations,

Reviewer: Replace this entire section with:
```suggestion
You can use Visual Studio Code (VS Code) extensions to set your environment, create launch configurations,
and browse and download samples.

The basic steps to build and run a sample using VS Code include:
 - Download a sample using the extension **Code Sample Browser for Intel&reg; oneAPI Toolkits**.
 - Configure the oneAPI environment with the extension **Environment Configurator for Intel&reg; oneAPI Toolkits**.
 - Open a Terminal in VS Code (**Terminal>New Terminal**).
 - Run the sample in the VS Code terminal using the instructions below.

To learn more about the extensions and how to configure the oneAPI environment, see the
[Using Visual Studio Code with Intel&reg;oneAPI Toolkits User Guide](https://software.intel.com/content/www/us/en/develop/documentation/using-vs-code-with-intel-oneapi/top.html).
```

and browse and download samples.

The basic steps to build and run a sample using VS Code include:
 - Download a sample using the extension **Code Sample Browser for Intel oneAPI Toolkits**.
 - Configure the oneAPI environment with the extension **Environment Configurator for Intel oneAPI Toolkits**.
 - Open a Terminal in VS Code (**Terminal>New Terminal**).
 - Run the sample in the VS Code terminal using the instructions below.

To learn more about the extensions and how to configure the oneAPI environment, see
[Using Visual Studio Code with Intel® oneAPI Toolkits](https://software.intel.com/content/www/us/en/develop/documentation/using-vs-code-with-intel-oneapi/top.html).

> After learning how to use the extensions for Intel oneAPI Toolkits, return to this readme for instructions on how to build and run a sample.

Reviewer: ```suggestion

```

### On a Linux* System

1. Generate the `Makefile` by running `cmake`.
     ```
   mkdir build
   cd build
   ```

>    To compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:

Reviewer: ```suggestion
   To compile for the Intel&reg;PAC with Intel Arria&reg;10 GX FPGA, run `cmake` using the command:
```

    ```
    cmake ..
   ```

>    Alternatively, to compile for the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX), run `cmake` using the command:

Reviewer: ```suggestion
   Alternatively, to compile for the Intel&reg;FPGA PAC D5005 (with Intel Stratix&reg;10 SX), run `cmake` using the command:
```

   ```
   cmake .. -DFPGA_BOARD=intel_s10sx_pac:pac_s10
   ```
   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   ```
   cmake .. -DFPGA_BOARD=<board-support-package>:<board-variant>
   ```

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   * Compile for emulation (fast compile time, targets emulated FPGA device):
      ```
      make fpga_emu
      ```
   * Generate the optimization report:
     ```
     make report
     ```
   * Compile for FPGA hardware (longer compile time, targets FPGA device):
     ```
     make fpga
     ```

### On a Windows* System

1. Generate the `Makefile` by running `cmake`.
     ```
   mkdir build
   cd build
   ```

>    To compile for the Intel® PAC with Intel Arria® 10 GX FPGA, run `cmake` using the command:

Reviewer: ```suggestion
   To compile for the Intel&reg;PAC with Intel Arria&reg;10 GX FPGA, run `cmake` using the command:
```

    ```
    cmake -G "NMake Makefiles" ..
   ```

>    Alternatively, to compile for the Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX), run `cmake` using the command:

Reviewer: ```suggestion
   Alternatively, to compile for the Intel&reg;FPGA PAC D5005 (with Intel Stratix&reg;10 SX), run `cmake` using the command:
```

   ```
   cmake -G "NMake Makefiles" .. -DFPGA_BOARD=intel_s10sx_pac:pac_s10
   ```
   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the command:
   ```
   cmake -G "NMake Makefiles" .. -DFPGA_BOARD=<board-support-package>:<board-variant>
   ```

2. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:

   * Compile for emulation (fast compile time, targets emulated FPGA device):
     ```
     nmake fpga_emu
     ```
   * Generate the optimization report:
     ```
     nmake report
     ```
   * Compile for FPGA hardware (longer compile time, targets FPGA device):
     ```
     nmake fpga
     ```

> *Note:* The Intel® PAC with Intel Arria® 10 GX FPGA and Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) do not support Windows*. Compiling to FPGA hardware on Windows* requires a third-party or custom Board Support Package (BSP) with Windows* support.

Reviewer: ```suggestion
> **Note**: The Intel&reg;PAC with Intel Arria&reg;10 GX FPGA and Intel&reg;FPGA PAC D5005
(with Intel Stratix&reg;10 SX) do not support Windows*. Compiling to FPGA hardware
on Windows* requires a third-party or custom Board Support Package (BSP) with
Windows* support.
```

> 

Reviewer: ```suggestion
> **Note**: If you encounter any issues with long paths when
compiling under Windows*, you may have to create your ‘build’ directory in a
shorter path, for example c:\samples\build. You can then run cmake from that
directory, and provide cmake with the full path to your sample directory.

### Troubleshooting
If an error occurs, you can get more details by running `make` with
the `VERBOSE=1` argument:
``make VERBOSE=1``
For more comprehensive troubleshooting, use the Diagnostics Utility for
Intel&reg;oneAPI Toolkits, which provides system checks to find missing
dependencies and permissions errors.
[Learn more](https://software.intel.com/content/www/us/en/develop/documentation/diagnostic-utility-user-guide/top.html).
```

 ### In Third-Party Integrated Development Environments (IDEs)

> You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide)

Reviewer: ```suggestion
You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [FPGA Workflows on Third-Party IDEs for Intel&reg;oneAPI Toolkits](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html)
```

## Running the Sample

 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
     ```
     ./mmhost.fpga_emu     (Linux)
     mmhost.fpga_emu.exe   (Windows)
     ```
2. Run the sample on the FPGA device:
     ```

>      ./mmhost.fpga         (Linux)

Reviewer: ```suggestion
     ./mmhost.fpga         (Linux)
     mmhost.fpga.exe     (Windows)
```
